packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -507,6 +465,105 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
       return throwBadResponse(stateP, res);
     }
+  } else if (stateP.controlState === 'CLEANUP_UNKNOWN_AND_EXCLUDED') {
+    const res = resW as ExcludeRetryableEsError<ResponseType<typeof stateP.controlState>>;
+    if (Either.isRight(res)) {
+      if (res.right.unknownDocs.length) {
+        logs = [
+          ...stateP.logs,
+          { level: 'warning', message: extractDiscardedUnknownDocs(res.right.unknownDocs) },
+        ];
+      }
+
+      logs = [
+        ...logs,
+        ...Object.entries(res.right.errorsByType).map(([soType, error]) => ({
+          level: 'warning' as const,
+          message: `Ignored excludeOnUpgrade hook on type [${soType}] that failed with error: "${error.toString()}"`,
+        })),
+      ];
+
+      return {
+        ...stateP,
+        logs,
+        controlState: 'CLEANUP_UNKNOWN_AND_EXCLUDED_WAIT_FOR_TASK',
+        deleteByQueryTaskId: res.right.taskId,
+      };
+    } else {
+      return {
+        ...stateP,
+        controlState: 'FATAL',
+        reason: extractUnknownDocFailureReason(
+          stateP.migrationDocLinks.resolveMigrationFailures,
+          res.left.unknownDocs
+        ),
+      };
+    }
+  } else if (stateP.controlState === 'CLEANUP_UNKNOWN_AND_EXCLUDED_WAIT_FOR_TASK') {
+    const res = resW as ExcludeRetryableEsError<ResponseType<typeof stateP.controlState>>;
+    if (Either.isRight(res)) {
+      const source = stateP.sourceIndex.value;
+      return {
+        ...stateP,
+        logs,
+        controlState: 'PREPARE_COMPATIBLE_MIGRATION',
+        targetIndexRawMappings: stateP.sourceIndexMappings,
+        targetIndexMappings: mergeMigrationMappingPropertyHashes(
+          stateP.targetIndexMappings,
+          stateP.sourceIndexMappings
+        ),
+        preTransformDocsActions: [
+          // Point the version alias to the source index. This let's other Kibana
+          // instances know that a migration for the current version is "done"
+          // even though we may be waiting for document transformations to finish.
+          { add: { index: source!, alias: stateP.versionAlias } },
+          ...buildRemoveAliasActions(source!, Object.keys(stateP.aliases), [
+            stateP.currentAlias,
+            stateP.versionAlias,
+          ]),
+        ],
+      };
+    } else {
+      if (isTypeof(res.left, 'wait_for_task_completion_timeout')) {
+        // After waiting for the specified timeout, the task has not yet
+        // completed. Retry this step to see if the task has completed after an
+        // exponential delay.  We will basically keep polling forever until the
+        // Elasticsearch task succeeds or fails.
+        return delayRetryState(stateP, res.left.message, Number.MAX_SAFE_INTEGER);
+      } else {
+        return {
+          ...stateP,
+          controlState: 'FATAL',
+          reason:
+            `Migration failed because it was unable to delete unwanted documents from the ${stateP.sourceIndex.value} system index:\n` +
```

Comment:
we should ignore version conflicts and proceed

Reply:
I first implemented a custom `retryDelay` mechanism here, to make sure the delete is re-run if there are failures.
Then I got rid of it, I thought that by failing with FATAL we'd restart and re-run the delete (checking again for unwanted documents).

If we ignore the failures and carry on with the migration, we're going to update the aliases on the next step.
Thus, if for some reason we fail to delete one of the SO with unknown type, Kibana is not going to be able to start, nor to delete that document at startup, right?

Reply:
yeah I should have thought more carefully through this before saying we should proceed...

I don't think we should throw a FATAL even if everything will be fine in the end after we restart. Upgrades are a sensitive procedure and it's alarming to a user to see `FATAL` in the logs. This also feels like something that could happen fairly frequently it's not just an obscure edge case.

So I think the behaviour you had where we retry would be better. Looking at the diff you used to retry 5 times but I think we can retry `retryAttempts` times which defaults to 15.